src/BindPattern.jl

```diff
@@ -540,3 +532,69 @@ function bind_expression(location::LineNumberNode, expr, assigned::ImmutableDict
 
     return BoundExpression(location, expr, new_assigned)
 end
+
+is_empty_block(x) = is_expr(x, :block) && all(a -> a isa LineNumberNode, x.args)
+
+#
+# Bind a case.
+#
+function bind_case(
+    case_number::Int,
+    location::LineNumberNode,
+    case,
+    predeclared_temps,
+    binder::BinderContext)::BoundCase
+    while true
+        # do some rewritings if needed
+        if is_expr(case, :macrocall)
+            # expand top-level macros only
+            case = macroexpand(binder.mod, case, recursive=false)
+
+        elseif is_expr(case, :tuple, 2) && is_case(case.args[2]) && is_expr(case.args[2].args[2], :if)
+            # rewrite `pattern, if guard end => result`, which parses as
+            # `pattern, (if guard end => result)`
+            # to `(pattern, if guard end) => result`
+            # so that the guard is part of the pattern.
+            pattern = case.args[1]
+            if_guard = case.args[2].args[2]
+            result = case.args[2].args[3]
+            case = :(($pattern, $if_guard) => $result)
+
+        elseif is_case(case)
+            # rewrite `(pattern, if guard end) => result`
+            # to `(pattern where guard) => result`
```

Comment:
We do it in two steps because it always parses the second (not quite right) way unless the user parenthesizes it the first way, which we ought to handle.